### PR TITLE
Term combinators respecting the canonical structure of branches and return predicate

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** architecture (on top of #7080)

This is a new step in the direction of CEP https://github.com/coq/ceps/pull/34.

This commit proposes some combinators which seem appropriate in the direction of seeing the spine of lambda and let-ins in the branches and return predicate of a `Case` node (i.e. Coq's `match`) as an integral part of the structure of a `Case`.

The current proposal is certainly not the final version for such combinators yet, but feedback is needed so I'm submitting them as they are now. In particular, I have the following questions:

- Do we continue to replicate `Constr` to `Term`, or do we progressively deprecate `Term`?
- Do we try to have a minimal number of functions in `Term` or do we prefer
  to share code between kernel (`Term`/`Constr`) and higher levels (`Termops`/`EConstr`)? For instance, do we share the code of `EConstr.map` and `Constr.map` (which is  not the case currently)?
- Also, do we put (effective code of) combinators in `EConstr` or in `Termops`, or do we put them in `EConstr` and replicate them in `Termops`?

PS: I introduced a `userview` which changes the semantics of what is a subterm of a `match`. For instance, the body of the return predicate, the body of a branch, the types and let's in the context of a return predicate or branch are now subterms of a `match`. Also, as a consequence, there is no more need for β for contracting the branches of a `Case` after a ɩ: this shall now be part of ɩ. On the other side a branch of the form `Lambda (name, type_of_pattern_variable, body)` is not anymore a subterm (on which `pattern` would apply). The use of a `userview` is transitional. The idea is that eventually only the `userview` is used.

### CODE REVIEW

**Review by gares — COMMENTED** [review]
I see many "open", are they intended?

**Comment by ejgallego** [comment]
@gares that's coming for the PR this PR depends on

**Comment by Zimmi48** [comment]
Why do you say this? You should look only at the second commit as the first one is from another PR. I don't see any `open` there.

**Comment by ejgallego** [comment]
> sorry for the noise, I did not understand Emilio's comment and I've missed the "ontop" comment

Our process for depending-PRs is still very primitive so it is normal we all get confused.

**Comment by herbelin** [comment]
I'd be glad to have comments on this PR, if someone is interested. And if not, I believe that it can be merged.

File: engine/eConstr.mli
Comment (gares) [inline engine/eConstr.mli]: Shouldn't then be in the `Unsafe` module?

Comment (gares) [reply engine/eConstr.mli]: sorry for the noise

**Review by maximedenes — APPROVED** [review]
I'll merge soon if no feedback.

**Review by ppedrot — APPROVED** [review]

**Comment by Zimmi48** [comment]
@ppedrot @maximedenes master is broken following the merge of this PR which ran with an outdated CI: the mli-doc need fixing.

**Comment by Zimmi48** [comment]
(Sorry @ppedrot I meant @herbelin)

**Comment by maximedenes** [comment]
@Zimmi48 am I expected to do something about it? If so, what should I do?

**Comment by Zimmi48** [comment]
@maximedenes I just wanted to let you know about it. Be more careful next time the merge script tells you the CI changed. I opened a PR with a fix at #8470.

**Comment by maximedenes** [comment]
> Be more careful next time the merge script tells you the CI changed

I don't think rebasing each time scales, so not sure what I can do about it.

**Comment by Zimmi48** [comment]
>I don't think rebasing each time scales, so not sure what I can do about it.

The script warns you specifically when the CI setup changes and this is not so common.

**Comment by maximedenes** [comment]
> this is not so common

Maybe there's a bug then, because I basically get the warning each time I merge a PR. Your PR fixing the `mli-doc` target was the first one not triggering the warning that I can remember of.

**Comment by Zimmi48** [comment]
That's because you take too long to merge the PRs you are in charge of :wink: I almost never see the warning.

**Comment by maximedenes** [comment]
Yeah indeed, I have other activities. But having to also rebase the PRs is only making the problem worse. I thought about removing myself from some maintainer group like doc, hoping to improving my reactivity.

**Comment by Zimmi48** [comment]
@maximedenes It would probably be a very good move.

**Pull request merged**
